internal/rukpak/source/unpacker.go

```diff
@@ -0,0 +1,132 @@
+package source
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/fs"
+	"net/http"
+
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	clusterextension "github.com/operator-framework/operator-controller/api/v1alpha1"
+	rukpakapi "github.com/operator-framework/operator-controller/internal/rukpak/api"
+)
+
+// Unpacker unpacks bundle content, either synchronously or asynchronously and
+// returns a Result, which conveys information about the progress of unpacking
+// the bundle content.
+//
+// If a Source unpacks content asynchronously, it should register one or more
+// watches with a controller to ensure that Bundles referencing this source
+// can be reconciled as progress updates are available.
+//
+// For asynchronous Sources, multiple calls to Unpack should be made until the
+// returned result includes state StateUnpacked.
+//
+// NOTE: A source is meant to be agnostic to specific bundle formats and
+// specifications. A source should treat a bundle root directory as an opaque
+// file tree and delegate bundle format concerns to bundle parsers.
+type Unpacker interface {
+	Unpack(context.Context, *rukpakapi.BundleSource, *clusterextension.ClusterExtension) (*Result, error)
+}
+
+// Result conveys progress information about unpacking bundle content.
+type Result struct {
+	// Bundle contains the full filesystem of a bundle's root directory.
+	Bundle fs.FS
+
+	// ResolvedSource is a reproducible view of a Bundle's Source.
+	// When possible, source implementations should return a ResolvedSource
+	// that pins the Source such that future fetches of the bundle content can
+	// be guaranteed to fetch the exact same bundle content as the original
+	// unpack.
+	//
+	// For example, resolved image sources should reference a container image
+	// digest rather than an image tag, and git sources should reference a
+	// commit hash rather than a branch or tag.
+	ResolvedSource *rukpakapi.BundleSource
```

Comment:
I'm wondering if we need the ResolvedSource here? What if we directly populate the ClusterExtension status with the details of resolved bundle?

Reply:
Also, I was trying to avoid changing k8s APIs